controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,890 @@
+	}
+	resourceClusterClient, err := newClustersClient(azureAuthorizer, spec)
+	if err != nil {
+		return nil, err
+	}
+	clusterState, err := resourceClusterClient.Get(ctx, spec.ResourceGroup, spec.ClusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	// set Kubernetes version
+	if clusterState.KubernetesVersion == nil {
+		return nil, fmt.Errorf("cannot detect cluster [%s] upstream kubernetes version", spec.ClusterName)
+	}
+	upstreamSpec.KubernetesVersion = clusterState.KubernetesVersion
+
+	// set tags
+	upstreamSpec.Tags = make(map[string]string)
+	if len(clusterState.Tags) != 0 {
+		upstreamSpec.Tags = to.StringMap(clusterState.Tags)
+	}
+
+	// set AgentPool profile
+	for _, np := range *clusterState.AgentPoolProfiles {
+		var upstreamNP aksv1.AKSNodePool
+		upstreamNP.Name = np.Name
+		upstreamNP.Count = np.Count
+		upstreamNP.MaxPods = np.MaxPods
+		upstreamNP.VMSize = string(np.VMSize)
+		upstreamNP.OsDiskSizeGB = np.OsDiskSizeGB
+		upstreamNP.OsDiskType = string(np.OsDiskType)
+		upstreamNP.Mode = string(np.Mode)
+		upstreamNP.OsType = string(np.OsType)
+		upstreamNP.OrchestratorVersion = np.OrchestratorVersion
+		upstreamNP.AvailabilityZones = np.AvailabilityZones
+		if np.EnableAutoScaling != nil {
+			upstreamNP.EnableAutoScaling = np.EnableAutoScaling
+			upstreamNP.MaxCount = np.MaxCount
+			upstreamNP.MinCount = np.MinCount
+		}
+		upstreamSpec.NodePools = append(upstreamSpec.NodePools, upstreamNP)
+	}
+
+	// set network configuration
+	networkProfile := clusterState.NetworkProfile
+	if networkProfile != nil {
+		upstreamSpec.NetworkPlugin = string(networkProfile.NetworkPlugin)
+		upstreamSpec.NetworkDNSServiceIP = to.String(networkProfile.DNSServiceIP)
+		upstreamSpec.NetworkDockerBridgeCIDR = to.String(networkProfile.DockerBridgeCidr)
+		upstreamSpec.NetworkServiceCIDR = to.String(networkProfile.ServiceCidr)
+		upstreamSpec.NetworkPolicy = string(networkProfile.NetworkPolicy)
+		upstreamSpec.NetworkPodCIDR = to.String(networkProfile.PodCidr)
+		upstreamSpec.LoadBalancerSku = string(networkProfile.LoadBalancerSku)
+	}
+
+	// set linux account profile
+	linuxProfile := clusterState.LinuxProfile
+	if linuxProfile != nil {
+		upstreamSpec.LinuxAdminUsername = to.String(linuxProfile.AdminUsername)
+		sshKeys := *linuxProfile.SSH.PublicKeys
+		upstreamSpec.LinuxSSHPublicKeyContents = to.String(sshKeys[0].KeyData)
+	}
+
+	// set API server access profile
+	apiServerAccessProfile := clusterState.APIServerAccessProfile
+	if to.Bool(apiServerAccessProfile.EnablePrivateCluster) {
+		upstreamSpec.PrivateCluster = apiServerAccessProfile.EnablePrivateCluster
+	}
+	if apiServerAccessProfile.AuthorizedIPRanges != nil {
+		upstreamSpec.AuthorizedIPRanges = apiServerAccessProfile.AuthorizedIPRanges
+	}
+
+	return upstreamSpec, err
+}
+
+func updateUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.SecretCache, spec aksv1.AKSClusterConfigSpec, upstreamSpec *aksv1.AKSClusterConfigSpec) (*aksv1.AKSClusterConfigSpec, error) {
```

Comment:
Nit: Would be nice to have comments on some of these eventually.

Reply:
Is this function not done yet ? Guessing you would fill out during syncing/refreshing phase of work ?

Reply:
This function is done. It compares the upstream spec with the config spec, then updates the upstream AKS cluster to match the config spec. Function returns after a update is finished.

Reply:
I was asking because there is a ton of specific logic in the EKS version: https://github.com/rancher/eks-operator/blob/master/controller/eks-cluster-config-handler.go#L907

So maybe the private and feature stuff is not in there yet and will be added later ?

Reply:
I _think_ the private cluster is implemented; not sure about the feature part - was that for GKE or AKS?

Reply:
Yes, `PrivateCluster` feature was added, we have to only test tunneling